src/kvstore.c

```diff
@@ -0,0 +1,749 @@
+     * As the addition and subtraction cancel each other out, we can simply return the result. */
+    return result;
+}
+
+/* Returns next non-empty dict index strictly after given one, or -1 if provided didx is the last one. */
+int kvstoreGetNextNonEmptyDictIndex(kvstore *kvs, int didx) {
+    unsigned long long next_key = cumulativeKeyCountRead(kvs, didx) + 1;
+    return next_key <= kvstoreSize(kvs) ? kvstoreFindDictIndexByKeyIndex(kvs, next_key) : -1;
+}
+
+int kvstoreNumNonEmptyDicts(kvstore *kvs) {
+    return kvs->non_empty_dicts;
+}
+
+int kvstoreNumDicts(kvstore *kvs) {
+    return kvs->num_dicts;
+}
+
+/* Returns kvstore iterator that can be used to iterate through sub-dictionaries.
+ *
+ * The caller should free the resulting kvs_it with kvstoreIteratorRelease. */
+kvstoreIterator *kvstoreIteratorInit(kvstore *kvs) {
+    kvstoreIterator *kvs_it = zmalloc(sizeof(*kvs_it));
+    kvs_it->kvs = kvs;
+    kvs_it->didx = -1;
+    kvs_it->next_didx = kvstoreFindDictIndexByKeyIndex(kvs_it->kvs, 1); /* Finds first non-empty dict index. */
+    dictInitSafeIterator(&kvs_it->di, NULL);
+    return kvs_it;
+}
+
+/* Free the dbit returned by dbIteratorInit. */
+void kvstoreIteratorRelease(kvstoreIterator *kvs_it) {
+    dictIterator *iter = &kvs_it->di;
+    dictResetIterator(iter);
+
+    zfree(kvs_it);
+}
+
+/* Returns next dictionary from the iterator, or NULL if iteration is complete. */
+dict *kvstoreIteratorNextDict(kvstoreIterator *kvs_it) {
+    if (kvs_it->next_didx == -1)
+        return NULL;
+    kvs_it->didx = kvs_it->next_didx;
+    kvs_it->next_didx = kvstoreGetNextNonEmptyDictIndex(kvs_it->kvs, kvs_it->didx);
+    return kvs_it->kvs->dicts[kvs_it->didx];
+}
+
+int kvstoreIteratorGetCurrentDictIndex(kvstoreIterator *kvs_it) {
+    assert(kvs_it->didx >= 0 && kvs_it->didx < kvs_it->kvs->num_dicts);
+    return kvs_it->didx;
+}
+
+/* Returns next entry. */
+dictEntry *kvstoreIteratorNext(kvstoreIterator *kvs_it) {
+    dictEntry *de = kvs_it->di.d ? dictNext(&kvs_it->di) : NULL;
+    if (!de) { /* No current dict or reached the end of the dictionary. */
+        dict *d = kvstoreIteratorNextDict(kvs_it);
+        if (!d)
+            return NULL;
+        if (kvs_it->di.d) {
+            /* Before we move to the next dict, reset the iter of the previous dict. */
+            dictIterator *iter = &kvs_it->di;
+            dictResetIterator(iter);
+        }
+        dictInitSafeIterator(&kvs_it->di, d);
+        de = dictNext(&kvs_it->di);
+    }
+    return de;
+}
+
+/* This method traverses through kvstore dictionaries and triggers a resize .
+ * It first tries to shrink if needed, and if it isn't, it tries to expand. */
+void kvstoreTryResizeDicts(kvstore *kvs, int limit) {
+    if (limit > kvs->num_dicts)
+        limit = kvs->num_dicts;
+
+    for (int i = 0; i < limit; i++) {
+        int didx = kvs->resize_cursor;
+        dict *d = kvstoreGetDict(kvs, didx);
+        if (!d)
+            continue;
```

Comment:
we keep falling on that test in daily because it is skipped in the CI cycle.
in retrospect, i should have run daily on such a big PR before merging it.